File: core/BestTimeToBuyAndSellStock/BestTimeToBuyAndSellStock.py
from typing import List

def maxProfit(prices: List[int]) -> int:
    buy, sell = -prices[0], 0
    res = sell
    for i in range(1, len(prices)):
        temp_buy = buy
        buy = max(buy, -prices[i])
        sell = max(sell, temp_buy + prices[i])
        res = max(res, sell)

    return res

File: core/BestTimeToBuyAndSellStock/BestTimeToBuyAndSellStock.java
package com.dbc;

public class BestTimeToBuyAndSellStock {
    public int maxProfit(int[] prices) {
        int buy = -prices[0], sell = 0;
        int res = sell;
        for (int i = 1; i < prices.length; i++) {
            sell = Math.max(sell, buy + prices[i]);
            buy = Math.max(buy, -prices[i]);
            res = Math.max(res, sell);
        }
        return res;
    }
}
